File: javascript/RationalNumbers.js
export class Rational {
  constructor(numerator, denominator) {
    this.numerator = numerator;
    this.denominator = denominator;
    // It's good practice to reduce in the constructor as well,
    // although the tests don't strictly require it for initial creation.
    // For this problem, we'll rely on explicit reduce calls in methods.
  }

  // Adds two rational numbers and reduces the result
  add(r) {
    return new Rational(
      this.numerator * r.denominator + r.numerator * this.denominator,
      this.denominator * r.denominator
    ).reduce();
  }

  // Subtracts two rational numbers and reduces the result
  sub(r) {
    return new Rational(
      this.numerator * r.denominator - r.numerator * this.denominator,
      this.denominator * r.denominator
    ).reduce();
  }

  // Multiplies two rational numbers and reduces the result
  mul(r) {
    return new Rational(
      this.numerator * r.numerator,
      this.denominator * r.denominator
    ).reduce();
  }

  // Divides two rational numbers and reduces the result
  div(r) {
    return new Rational(
      this.numerator * r.denominator,
      r.numerator * this.denominator
    ).reduce();
  }

  // Calculates the absolute value of the rational number and reduces the result
  abs() {
    // Corrected: Call .reduce() after creating the new Rational object
    return new Rational(Math.abs(this.numerator), Math.abs(this.denominator)).reduce();
  }

  // Reduces the rational number to its lowest terms
  reduce() {
    // Handle the case where the numerator is zero
    if (this.numerator === 0) {
      return new Rational(0, 1);
    }

    // Calculate the greatest common divisor (GCD) of the absolute values
    const commonDivisor = this.gcd(
      Math.abs(this.numerator),
      Math.abs(this.denominator)
    );

    // Determine the sign of the result.
    // If numerator and denominator have different signs, the result is negative.
    // Otherwise, it's positive.
    const sign = (this.numerator * this.denominator < 0) ? -1 : 1;

    // Calculate the reduced numerator and denominator
    const reducedNumerator = Math.abs(this.numerator) / commonDivisor;
    const reducedDenominator = Math.abs(this.denominator) / commonDivisor;

    // Return a new Rational with the correct sign applied to the numerator
    return new Rational(sign * reducedNumerator, reducedDenominator);
  }

  // Raises the rational number to an integer power
  exprational(power) {
    if (power === 0) {
      return new Rational(1, 1); // Any non-zero rational number to the power of 0 is 1
    }
    if (this.numerator === 0 && power < 0) {
      throw new Error("Cannot raise zero to a negative power"); // Handle division by zero
    }

    return power >= 0
      ? new Rational(
          Math.pow(this.numerator, power),
          Math.pow(this.denominator, power)
        ).reduce()
      : new Rational(
          Math.pow(this.denominator, Math.abs(power)), // Invert for negative power
          Math.pow(this.numerator, Math.abs(power))
        ).reduce();
  }

  // Raises a real number to the power of this rational number
  expreal(baseNumber) {
    const result = Math.pow(baseNumber, this.numerator / this.denominator);
    // Define a small tolerance for floating-point comparisons
    const epsilon = 1e-9; // A common value for floating-point precision

    // If the result is very close to an integer, return the integer
    if (Math.abs(result - Math.round(result)) < epsilon) {
      return Math.round(result);
    }

    // Otherwise, return the result as is
    return result;
  }

  // Helper function to calculate the greatest common divisor (GCD) using Euclidean algorithm
  gcd(a, b) {
    // Ensure a and b are non-negative for GCD calculation
    a = Math.abs(a);
    b = Math.abs(b);
    while (b) {
      [a, b] = [b, a % b];
    }
    return a;
  }
}

File: c/RationalNumbers.c
#include "rational_numbers.h"
int gcd(int a, int b)
{
    register int aux;
    if(a < b)
    {
        aux = a;
        a = b;
        b = aux;
    }
    if(a < 0) a = -a;
    if(b < 0) b = -b;
    if(!b) return a; // gcd(a, 0) = a
    if(b == 1) return 1; // gcd(a, 1) = 1
    if(!(a & 1) && !(b & 1))
        return gcd(a >> 1, b >> 1) << 1; // gcd(2u, 2v) = 2 gcd(u, v)
    // gcd(2u, v) = gcd(u, 2v) = gcd(u, v)
    if((!(a & 1) && (b & 1)))
        return gcd(a >> 1, b);
    if(((a & 1) && !(b & 1)))
        return gcd(a, b >> 1);
    // gcd(a, b) = gcd(a - b, b)
    return gcd(a - b, b);
}
float root(float r, int n)
{
    int i, j;
    float x, xp;
    x = xp = 1;
    for(i = 0; i < 5; ++i)
    {
        for(xp = 1, j = n - 1; j; --j) xp *= x;
        x = (n - 1) * x / n + r / (n * xp);
    }
    return x;
}
rational_t reduce(rational_t r)
{
    int g = gcd(r.numerator, r.denominator);
    r.numerator /= g;
    r.denominator /= g;
    if(r.denominator < 0)
    {
        r.denominator = -r.denominator;
        r.numerator = -r.numerator;
    }
    return r;
}
rational_t add(rational_t r1, rational_t r2)
{
    rational_t r;
    r.numerator = r1.numerator * r2.denominator + r2.numerator * r1.denominator;
    r.denominator = r1.denominator * r2.denominator;
    return reduce(r);
}
rational_t subtract(rational_t r1, rational_t r2)
{
    rational_t r;
    r.numerator = r1.numerator * r2.denominator - r2.numerator * r1.denominator;
    r.denominator = r1.denominator * r2.denominator;
    return reduce(r);
}
rational_t multiply(rational_t r1, rational_t r2)
{
    rational_t r;
    r.numerator = r1.numerator * r2.numerator;
    r.denominator = r1.denominator * r2.denominator;
    return reduce(r);
}
rational_t divide(rational_t r1, rational_t r2)
{
    rational_t r;
    r.numerator = r1.numerator * r2.denominator;
    r.denominator = r1.denominator * r2.numerator;
    return reduce(r);
}
rational_t absolute(rational_t r)
{
    r = reduce(r);
    if(r.numerator < 0) r.numerator = -r.numerator;
    return r;
}
rational_t exp_rational(rational_t b, int e)
{
    rational_t r = {1, 1};
    int aux;
    if(!e) return r;
    if(e < 0)
    {
        e = -e;
        aux = b.numerator;
        b.numerator = b.denominator;
        b.denominator = aux;
    }
    while(e--)
    {
        r.numerator *= b.numerator;
        r.denominator *= b.denominator;
    }
    return reduce(r);
}
float exp_real(float b, rational_t e)
{
    rational_t abs_e;
    float r = 1, r_;
    if(!e.numerator) return 1.;
    e = reduce(e);
    abs_e = absolute(e);
    r_ = root(b, e.denominator);
    while(abs_e.numerator--) r *= r_;
    if(e.numerator < 0) return 1.0/r;
    else        return r;
}

/*
rational_numbers.h
#ifndef RATIONAL_NUMBERS_H
#define RATIONAL_NUMBERS_H
typedef struct
{
    int numerator;
    int denominator;
}
rational_t;
int gcd(int a, int b);
float root(float r, int n);
rational_t reduce(rational_t r);
rational_t add(rational_t r1, rational_t r2);
rational_t subtract(rational_t r1, rational_t r2);
rational_t multiply(rational_t r1, rational_t r2);
rational_t divide(rational_t r1, rational_t r2);
rational_t absolute(rational_t r);
rational_t exp_rational(rational_t b, int e);
float exp_real(float b, rational_t e);
#endif
*/